feat: add epoching module (#10132)
cosmos/cosmos-sdk: x/epoching/spec/01_state.md

<!--
order: 1
-->

# State

## Messages queue

> Messages are queued to run at the end of epochs. Queued messages have epoch number to be run and at the end of epochs, it run messages queued for the epoch and execute the message.

Reviewer: ```suggestion
Messages are queued to run at the end of each epoch. Queued messages have an epoch number and for each epoch number, the queues are iterated over and each message is executed.
```

Reviewer: Also question, how do order iterating over module queues? Are there cases where this matters?

### Message queues

> Each module has 1 message queue that is specific to a module.

Reviewer: ```suggestion
Each module has one unique message queue that is specific to that module.
```

## Actions

> A module will add a message that complies with the `sdk.Msg` interface. These message will be executed at a later date.

Reviewer: ```suggestion
A module will add a message that implements the `sdk.Msg` interface. These message will be executed at a later time (end of the next epoch).
```

```go
type Msg interface {
  proto.Message

  // Return the message type.
  // Must be alphanumeric or empty.
  Route() string

  // Returns a human-readable string for the message, intended for utilization
  // within tags
  Type() string

  // ValidateBasic does a simple validation check that
  // doesn't require access to any other information.
  ValidateBasic() error

  // Get the canonical byte representation of the Msg.
  GetSignBytes() []byte

  // Signers returns the addrs of signers that must sign.
  // CONTRACT: All signatures must be present to be valid.
  // CONTRACT: Returns addrs in some deterministic order.
  GetSigners() []AccAddress
 }
```

## Buffered Messages Export / Import

> For now, it's implemented to export all buffered messages without epoch number. And when import, Buffered messages are stored on current epoch to run at the end of current epoch.

Reviewer: ```suggestion
For now, the `x/epoching` module is implemented to export all buffered messages without epoch numbers. When state is imported, buffered messages are stored on current epoch to run at the end of current epoch.
```
